### PR TITLE
Begin UI Refactor with Responsiveness & New Menus

### PR DESCRIPTION
**Pull request by danfinlay**
Ready for QA! Needs thorough QA to make sure all features are still accessible, useful, and their formatting is not messed up.

Rather than let this `NewUI` branch run any longer, we've made changes to it so it can be "production ready" asap.

This PR represents our first suspicion that *this may be ready to go to production*!

Also, it looks good in different view sizes, from desktop tabs to mobile.

Also includes the new work by Kyokan Labs, where they implemented the new drop-down designs by Christian, so this is starting to inch towards our new design template.

I'll keep track of outstanding changes before we can merge this here:

- [x] Custom RPC items remain checked and clicking them is ineffective.
- [x] Selecting `Localhost 8545` ineffective.

### CODE REVIEW

**Comment by flyswatter** [comment]
Okay, fixed the one bug I was aware of, @tmashuang this is ready for a QA pass when you have time.

File: docs/responsive-ui-dev.md
Comment (sdtsui) [inline docs/responsive-ui-dev.md]: Are these instructions still applicable now that the folders are merged?  Suggesting that we remove, or add some context about responsive development instead.

File: package.json
Comment (sdtsui) [inline package.json]: Might be low priority: could this be `2.0.1` since it's merged in `menu-droppo`?

File: ui/app/add-token.js
Comment (sdtsui) [inline ui/app/add-token.js]: typo?  (s*ym*bol..)

File: ui/app/account-detail.js
Comment (sdtsui) [inline ui/app/account-detail.js]: :+1:

**Review by sdtsui — COMMENTED** [review]
Minor suggestions.

**Comment by flyswatter** [comment]
Good catches, @sdtsui, implemented all those suggestions.

**Comment by tmashuang** [comment]
- [ ] In info.js, the Twitter link should be on a new line. Current CSS display is set to inline-block from fa, font-awesome, ```display: block``` should fix this for twitter link.
- [ ] Instance of Loading Spinner persisting after sending transactions. 
- [ ] Loading Spinner does not properly center
Note: When the extension URL is open in a window no popup appears when interaction with Dapps. Don't know if this is the intended functionality.

**Comment by flyswatter** [comment]
@tmashuang Thanks for the QA, I fixed 2/4 of them. On the remaining issues:

> Instance of Loading Spinner persisting after sending transactions.

I briefly saw this, it was related to a nonce not being an integer, which was a really weird bug, because we already adjusted provider engine to allow non-integer nonces. This then went away, and I haven't seen it again. Can you get this again? I think it may be fixed.

> Note: When the extension URL is open in a window no popup appears when interaction with Dapps. Don't know if this is the intended functionality.

This is because we have some logic that says "If the UI is already open when a TX is proposed, don't open another one". We'll have to revise that logic for Mascara, but for now, it's consistent with how the extension UI behaves correctly.

So anyways, if you could re-check that spinner-post-tx bug, and let me know if it's fixed, and if there's anything else, I think we're getting close.

File: ui/app/components/pending-tx.js
Comment (tmashuang) [inline ui/app/components/pending-tx.js]: Are we considering going back a minimal gas price of 2? #1796

Comment (flyswatter) [inline ui/app/components/pending-tx.js]: No, sorry, when I copied the `responsive-ui` folder over `ui` the merge seems to be prioritizing the `NewUI` over `master`, so it's changing some things like this.  Good catch, I'll review the diffs for other things like this.

Comment (flyswatter) [inline ui/app/components/pending-tx.js]: Yeah I've read over the diffs again, I didn't see any other changes we missed.

File: ui/app/components/token-list.js
Comment (tmashuang) [inline ui/app/components/token-list.js]: Possibility for only 1 token would display "You own 1 tokens"

**Comment by tmashuang** [comment]
Last one, in recover-seed/confirmation.js the width is for flex-space-between is rather large, 50%, could be reduced to 20.

**Comment by tmashuang** [comment]
NewUI in FF
newui_in_ff

**Comment by frankiebee** [comment]
I seem to be getting something slightly different in fullscreen in mascara am i missing some new build step?

screen shot 2017-07-27 at 7 34 07 pm

**Comment by flyswatter** [comment]
@frankiebee That's how it's supposed to work. It's flexy. Try resizing the window. It's approaching the layout that Christian has passed us.

**Comment by flyswatter** [comment]
Added a max-width per @tmashuang's advice. https://github.com/MetaMask/metamask-extension/pull/1835/files#diff-05f3f3e6a668f4f66e3c66bd417dd669R27

**Comment by flyswatter** [comment]
Interesting that Firefox automatically flexes the action-view wide, and Chrome doesn't. We might want to simply have hard-coded dimensions on the extension's html vs mascara or mobile.

**Comment by flyswatter** [comment]
Firefox does not have good style inspecting of its WebExtension pop-ups. I can't even look at which CSS property is forcing the view wide like that. I've contacted them on IRC (which has been one of my most responsive channels with them in the past).

Options:
- Hard coding popup dimensions, using a 2nd html file for responsiveness
- Accepting that Firefox has strange dimensions when it's used

**Comment by flyswatter** [comment]
Whoops didn't mean to close.

**Comment by flyswatter** [comment]
Okay, we now have two UI html files. `popup.html` includes inline styles to force it to the traditional size, like this:

screen shot 2017-07-28 at 11 07 50 am

The new `home.html` file is the one that has no fixed sizes, and supports a fully responsive interface.

**Comment by sdtsui** [comment]
fyi for @flyswatter - I noticed in both popup screenshots (Thomas' firefox one, and above): we're losing a few px off the bottom of the account address. Worth fixing?
(Doesn't happen with Frankie's responsive screenshot.)

**Comment by sdtsui** [comment]
Noticed this on latest. Global handler in `menu-droppo 2.0.1` appears to be having unintended side effects on dropdown behavior. Fixing now. 

2 0 1 effects

**Comment by sdtsui** [comment]
@flyswatter @tmashuang - PRing on top of this one to fix an issue I spotted.
#1841

**Comment by sdtsui** [comment]
I was preparing some docs for the state atom to make the redesign work easier.
Along the way, I was able to repro the post-tx-spinner issue in popup. Sharing here. 
```
651fec5 - (origin/NewUI, metamask/NewUI) Create distinct responsive 'home.html' file, hard-code popup.html size (8 hours ago) <Dan Finlay>hours ago) <Dan Finlay>
```

post-tx-spinner


Hope that helps.

**Comment by sdtsui** [comment]
Adding a bit more context, from the state atom:


Full Gist:
https://gist.github.com/sdtsui/65a03c1f6eb4257bb2cdf16da4e1ddaa

**Comment by flyswatter** [comment]
> we're losing a few px off the bottom of the account address. Worth fixing?

It's only on Firefox, so it's worth fixing for FF, but also, FF hasn't made it easy to analyze CSS for action-views like this, so it's hard to prioritize very seriously. They need to bring their extension debugging up a notch for anyone to bring really high quality extensions.

To reiterate something I said in our dev channel:

I think it'd be smart to deploy our current master before merging this, so we get more QA time on it. Our current master has some useful blacklist enhancements, while this style bump is not exactly urgent, and would benefit from more developer-usage time, we'll probably work through some kinks like that spinner bug.

I'm out for a week now, so good luck everyone on that bug and others!  Thanks for holding down the fort, good luck!

**Comment by sdtsui** [comment]
:+1: , thanks @flyswatter .  Enjoy the week off!

@Zanibas - shoot me a message on Gitter when you get this, we can team on next steps early Monday.  It'd be helpful to understand more about the deployment process as well.

**Comment by sdtsui** [comment]
tx-spinner-issue-on-newui-with-txdata

**Comment by sdtsui** [comment]
tx-spinner-issue-on-newui-with-txdata-2

**Comment by sdtsui** [comment]
tx repro gifs we recorded live:  cc for @Zanibas

**Comment by kumavis** [comment]
failing on lint - theres some unwated semicolons in there

**Comment by kumavis** [comment]
I think we lost a couple features in this PR
1) QR code display
2) copy-to-clipboard support

**Comment by kumavis** [comment]
I think we lost the QR code display in this PR

@sdtsui @Zanibas

**Comment by frankiebee** [comment]
when the window is at 797x781 the history and tokens snap to different sides
screen shot 2017-08-01 at 5 49 16 pm
screen shot 2017-08-01 at 5 49 22 pm

**Comment by kumavis** [comment]
@frankiebee thats intentional, its working towards the new design
it should prolly start at a thinner width though

**Comment by kumavis** [comment]
a lot of the targets feel really small on mobile (dropdown caret/dots, dropdown options)

**Comment by sdtsui** [comment]
Thanks @frankiebee @kumavis, we'll batch with the design call and review tomorrow. And thanks for the lint fix!

I wasn't aware QR code was staying: impression I got from Dan was to implement prototypes and remove all of the old buttons.  I'll add back with some sensible text for the option, but we should probably sync on any expected functionality losses for `newui` before I get farther along with it.

For example, newui doesn't have a QR code screen or a design for the 'custom RPC' option in network drop-down.

My instructions were to skip send token screen as well, and leave out anything without screens. Thoughts?

Finally, do either of you have thoughts on the tx spinner issue? Aside from some minor tweaks, that appears to be the main thing holding this PR back. Cheers.

**Comment by Zanibas** [comment]
Exporting a private key needs to be fixed (overflowing off-screen)

**Comment by sdtsui** [comment]
- Synced on above in design call. Thanks for clarifying, @kumavis .
- The issue raised by @tmashuang has been reproduced in #1853 .

**Comment by Zanibas** [comment]
Some scrollbars are not wide enough. See notices and their differing sidebars.
The sizing is inconsistent at first-time

**Comment by Zanibas** [comment]
We're going to stay with the text sizing on the droppos for now until later--not a blocking issue. We'll fix before anyone touches mobile, I promise :wink: 

We have to wait for the adjustment to the new font anyways to make semi-permanent calls on this.

**Comment by kumavis** [comment]
i wouldnt consider it blocking - but i do frequently use mascara on mobile, which is currently using the new ui

**Comment by Zanibas** [comment]
Okay, we'll elevate them then scale down if necessary.

**Comment by kumavis** [comment]
doesn't like the spread operator
https://github.com/MetaMask/metamask-extension/blob/5a9257208c858d11b94659518ef2375384aeac46/ui/app/components/dropdown.js#L76

**Comment by kumavis** [comment]
specifically `nyc`'s `istanbul` (coverage) does not like it

**Comment by kumavis** [comment]
i just noticed we had a [silent ci failure](https://circleci.com/gh/MetaMask/metamask-extension/2682) at 24a13d116ce4

```
Integration build failure { Error: ENOENT: no such file or directory, unlink '/home/ubuntu/metamask-extension/test/integration/bundle.js'
    at Object.fs.unlinkSync (fs.js:1066:18)
    at Object.<anonymous> (/home/ubuntu/metamask-extension/test/integration/index.js:11:6)
    at Module._compile (module.js:569:30)
    at Object.Module._extensions..js (module.js:580:10)
    at Module.load (module.js:503:32)
    at tryModuleLoad (module.js:466:12)
    at Function.Module._load (module.js:458:3)
    at Function.Module.runMain (module.js:605:10)
    at startup (bootstrap_node.js:158:16)
    at bootstrap_node.js:575:3
  errno: -2,
  code: 'ENOENT',
  syscall: 'unlink',
  path: '/home/ubuntu/metamask-extension/test/integration/bundle.js' }
```

**Comment by kumavis** [comment]
ok now it doesnt fail silently -- goood

**Comment by kumavis** [comment]
some sort of fatal integration test?

**Comment by kumavis** [comment]
thats right - our integration tests either pass or hang

**Comment by sdtsui** [comment]
@Zanibas : #1862 re-enables dropdown css transitions, as discussed this morning.  Please see that PR for details.

**Comment by sdtsui** [comment]
#1863 adds early breakpoint and resolved snapping weirdness.

**Comment by sdtsui** [comment]
@kumavis  thanks for merging. Looks like menu-droppo needs to be linted. Checking it out.

**Comment by sdtsui** [comment]
@kumavis /\

**Comment by sdtsui** [comment]
.. /\ looks like more silent failures.
```
> metamask-crx@0.0.0 genStates /home/ubuntu/metamask-extension
> node development/genStates.js
```
It'll wait 10m and probably time out.

**Comment by kumavis** [comment]
our integration tests either pass or hang on failure

**Comment by kumavis** [comment]
@sdtsui you can push directly to NewUI if you like (invited you to the repo https://github.com/MetaMask/metamask-extension/invitations)

**Comment by kumavis** [comment]
tests pass! awesome

**Comment by kumavis** [comment]
will review and merge it all soon, maybe tomorrow

**Review by sdtsui — COMMENTED** [review]
Cool, thanks - I'll respond to any changes on Monday.

**Review by kumavis — APPROVED** [review]

**Comment by kumavis** [comment]
This is beautiful 👌

**Pull request merged**
